File: code/backend/java/core/src/main/java/games/paths/core/port/match/CharacterQueryPort.java
package games.paths.core.port.match;

import games.paths.core.model.match.CharacterInstanceInfo;

import java.util.List;

/**
 * CharacterQueryPort - Inbound port for the read-side character operations.
 * Step 21: list the players of a match and read a single character detail.
 */
public interface CharacterQueryPort {

    /**
     * Lists the characters of a match. Returns {@code null} when the match does
     * not exist or the user is neither its creator nor a participant.
     */
    List<CharacterInstanceInfo> listPlayers(String matchUuid, String userUuid);

    /**
     * Returns the full detail of a single character in a match, or {@code null}
     * when the match/character does not exist or the user cannot access it.
     */
    CharacterInstanceInfo getCharacter(String matchUuid, String characterUuid, String userUuid);
}

File: code/backend/php/src/Core/Port/Matches/CharacterQueryPort.php
<?php

namespace Games\Paths\Core\Port\Matches;

use Games\Paths\Core\Domain\Matches\CharacterInstanceInfo;

interface CharacterQueryPort
{
    /**
     * Step 21 — list the characters of a match, or null when the match is
     * missing or the user has no access.
     *
     * @return CharacterInstanceInfo[]|null
     */
    public function listPlayers(string $matchUuid, string $userUuid): ?array;

    /**
     * Step 21 — return a single character detail, or null when missing / not
     * accessible.
     */
    public function getCharacter(string $matchUuid, string $characterUuid, string $userUuid): ?CharacterInstanceInfo;
}
